chatgpt_wrapper/chatgpt.py

```diff
@@ -333,48 +411,222 @@ def ask_stream(self, prompt: str):
                         full_event_message = "\n".join(
                             event["message"]["content"]["parts"]
                         )
-            except Exception:
-                yield (
+            except Exception as err:
+                logging.error("Failed to read response from ChatGPT: %s",err)
+                raise error.ChatGPTResponseError(
                     "Failed to read response from ChatGPT.  Tips:\n"
                     " * Try again.  ChatGPT can be flaky.\n"
                     " * Use the `session` command to refresh your session, and then try again.\n"
                     " * Restart the program in the `install` mode and make sure you are logged in."
                 )
-                break
 
             if full_event_message is not None:
                 chunk = full_event_message[len(last_event_msg):]
                 last_event_msg = full_event_message
+                self.log.info("< (ChatGPT) %s",chunk)
                 yield chunk
 
             # if we saw the eof signal, this was the last event we
             # should process and we are done
-            if len(eof_datas) > 0 or (((time.time() - start_time) > self.timeout) and full_event_message is None):
+            if len(eof_datas)>0:
                 break
 
-            sleep(0.2)
+            time.sleep(0.2)
+        else:
+            # Timeout
+            self.log.error("Timeout when asking ChatGPT. Retrying...")
+            self.refresh_session()
+            for chunk in self.ask_stream(prompt=prompt,remaining_retry=remaining_retry-1):
+                yield chunk
+            return
+
+        self._cleanup_divs()
+
+    def ask_stream_clicking(self, prompt: str, remaining_retry=2):
+        """Ask ChatGPT by clicking the buttons on the website. Response is streamed.
```

Comment:
What is the purpose of this function versus the other streaming function?

Reply:
It asks ChatGPT by clicking the button on the website. 
- I doubted if the `text-davinci-002-render-sha` was actually the one used in ChatGPT. Through the `network` pane of browser developer tools I've seen different models (although also this one). Therefore, to ensure a total ChatGPT experience I wrote this. 
- I hoped it to be more stable than the XHR way, but it turns out it didn't help. 

However, as you're adding new model options and `ask_stream_clicking` hasn't implemented conversation choosing, I think I can remove it now.

Reply:
Removed.